csm_web/scheduler/views.py

```diff
@@ -1,267 +1,150 @@
+        student = get_object_or_error(self.get_queryset(), pk=pk)
+        if request.method == 'GET':
+            return Response(AttendanceSerializer(student.attendance_set.all(), many=True).data)
+        # PUT
+        if student.user == self.request.user:
+            raise PermissionDenied('You cannot record your own attendance (nice try)')
+        try:  # update
+            attendance = student.attendance_set.get(date=request.data['date'])
+            serializer = AttendanceSerializer(attendance, data={'student': student.id, **request.data})
+        except ObjectDoesNotExist:  # create
+            serializer = AttendanceSerializer(data={'student': student.id, **request.data})
+        if serializer.is_valid():
+            serializer.save()
+            return Response(status=status.HTTP_204_NO_CONTENT)
+        return Response(serializer.errors, status=status.HTTP_422_UNPROCESSABLE_ENTITY)
 
 
-class OverrideDetail(generics.RetrieveUpdateDestroyAPIView):
-    queryset = Override.objects.all()
-    serializer_class = OverrideSerializer
-    permission_classes = (permissions.IsAuthenticatedOrReadOnly, IsLeaderOrReadOnly)
+class ProfileViewSet(*viewset_with('list')):
+    queryset = None
+    serializer_class = None
 
-    def get_object(self):
-        obj = get_object_or_404(self.get_queryset(), pk=self.kwargs["pk"])
-        self.check_object_permissions(self.request, obj)
-        return obj
+    def list(self, request):
+        student_profiles = StudentSerializer(request.user.student_set.all(), many=True).data
+        mentor_profiles = MentorSerializer(request.user.mentor_set.all(), many=True).data
+        return Response({'mentor_profiles': mentor_profiles, 'student_profiles': student_profiles})
 
 
-class CreateAttendanceDetail(generics.CreateAPIView):
-    queryset = Attendance.objects.all()
-    serializer_class = AttendanceSerializer
-    permission_classes = (permissions.IsAuthenticatedOrReadOnly, IsLeader)
+class SpacetimeViewSet(viewsets.GenericViewSet):
+    serializer_class = Spacetime
 
-    def perform_create(self, serializer):
-        # Deny attendance creation if not leader of section
-        profile = serializer.validated_data["attendee"]
-        section = profile.section
-        if not is_leader(self.request.user, section):
-            raise PermissionDenied(
-                "You are not allowed to create Attendances for that section"
-            )
-        elif not profile.active:  # Deny attendances for deactivated profiles
-            raise PermissionDenied(
-                "This profile ({}) has been deactivated".format(profile)
-            )
-        else:
+    def get_queryset(self):
+        return Spacetime.objects.filter(section__mentor__user=self.request.user)
+
+    @action(detail=True, methods=['put'])
+    def override(self, request, pk=None):
+        spacetime = get_object_or_error(self.get_queryset(), pk=pk)
+        if hasattr(spacetime, "override"):  # update
```

Comment:
I feel like there should be some kind of abstraction in models for this: perhaps add `has_any_override` to Spacetime?

Reply:
We do have `has_valid_override`, but here I'm specifically testing to see that an override exists, even if it's expired.

Reply:
Yeah, but I don't like breaking the abstraction barrier with hasattr here.